crates/sema/src/typeck/mod.rs

```diff
@@ -13,13 +17,48 @@ pub(crate) fn check(gcx: Gcx<'_>) {
         gcx.hir.par_contract_ids().for_each(|id| {
             check_duplicate_definitions(gcx, &gcx.symbol_resolver.contract_scopes[id]);
             check_payable_fallback_without_receive(gcx, id);
+            check_external_type_clashes(gcx, id);
         }),
         gcx.hir.par_source_ids().for_each(|id| {
             check_duplicate_definitions(gcx, &gcx.symbol_resolver.source_scopes[id]);
         }),
     );
 }
 
+fn check_external_type_clashes(gcx: Gcx<'_>, contract_id: hir::ContractId) {
+    if gcx.hir.contract(contract_id).kind.is_library() {
+        return;
+    }
+
+    let mut external_declarations: FxHashMap<U256, Vec<ItemId>> = FxHashMap::default();
+
+    for item_id in gcx.hir.contract_item_ids(contract_id) {
+        match gcx.hir.item(item_id) {
+            Item::Function(f) if f.is_part_of_external_interface() => {
+                let s = keccak256(gcx.item_signature(item_id));
```

Comment:
You'll need to make it pub(crate)

```suggestion
                let s = gcx.item_selector(item_id);
```

Reply:
done [68f23ef](https://github.com/paradigmxyz/solar/pull/312/commits/68f23efa440a33b4eedd90e35a18a345847879db) (limiting vis to crate is a good call)

we don't necessarily need to use keccak here since the key is not used except for collecting; wondering if just FxHasher is more performant (will hashing twice, once to find bucket in the map), or we can tell the map to drop key references (of `item_signature`) since we will never iterate over them

Reply:
yeah but this is fine since we already calculated all the selectors for external functions so we're just reusing the hash, and rehashing is very cheap